### PR TITLE
Allow executing dynamic tests/test template invocations by unique ID

### LINKED ISSUE #1025: Cannot select @TestTemplate or @TestFactory invocations by Unique ID
## Overview

With reference to [bug request 520923](https://bugs.eclipse.org/bugs/show_bug.cgi?id=520923) in Eclipse, I am trying to rerun a single invocation of a parameterized test (and similarly, a repeated and a dynamic test) using its unique ID like this:

``` java
LauncherDiscoveryRequest request = LauncherDiscoveryRequestBuilder.request().selectors(DiscoverySelectors.selectUniqueId(uniqueId)).build();
```

A new `Launcher` instance is created to discover the above request.

This method works for a normal test with ID:

``` java
[engine:junit-jupiter]/[class:test.FirstTest]/[method:myFirstTest(org.junit.jupiter.api.TestInfo)]
``` 

But it gives this error for a dynamic test:

```
org.junit.jupiter.engine.discovery.JavaElementsResolver resolveUniqueId
WARNING: Unique ID '[engine:junit-jupiter]/[class:test.dynamic.DynamicTestsDemo]/[test-factory:dynamicTestsFromIntStream()]/[dynamic-test:#2]' could not be resolved
```

or a parameterized test:

```
org.junit.jupiter.engine.discovery.JavaElementsResolver resolveUniqueId
WARNING: Unique ID '[engine:junit-jupiter]/[class:jb.Junit5SimpleParamTest]/[test-template:dummy(java.lang.String)]/[test-template-invocation:#1]' could not be resolved
```

In JUnit 4, it was done using `org.junit.runner.Request.filterWith(Filter filter)`. 

## Question

What is the recommended way to do this in JUnit 5?

## Related Issues

- #1026 
- #1031

## Analysis

Given a _Unique ID_ such as `[engine:junit-jupiter]/[class:test.dynamic.DynamicTestsDemo]/[test-factory:dynamicTestsFromIntStream()]/[dynamic-test:#2]`,
`JavaElementsResolver.resolveUniqueId(TestDescriptor, List<Segment>)` returns `true` if **all** segments of the supplied Unique ID could be resolved.

So, in such a case, the test class and test factory method do in fact get resolved, but the dynamic test itself does not get resolved.

That means that the test class and test factory method will be executed in the test plan even though there is a warning (which actually only applies to the last segment of the Unique ID).

## Proposals

1. Selection of any invocation of a test template or test factory results in selection of the test template or test factory as a _container_ and consequently the _selection_ of all invocations.
   - This appears to already be the case: see _Analysis_ for details.
2. Selection of a specific invocation of a test template or test factory results in selection of only that specific invocation of the test template or test factory. 
   - This could potentially be achieved via a special type of _filter_ that is applied to the stream of dynamic tests or test template invocations.

## Deliverables

- [ ] Ensure that an invocation of a `@TestTemplate` method can be selected by _Unique ID_.
  - This applies to `@RepeatedTest` and `@ParameterizedTest` as well as any _test template_ method.
- [ ] Ensure that an invocation of a dynamic test supplied by a `@TestFactory` method can be selected by _Unique ID_.

### PR DESCRIPTION
**Pull request by marcphilipp**
Resolves #1025.

---

I hereby agree to the terms of the JUnit Contributor License Agreement.

---

### Definition of Done

- [X] There are no TODOs left in the code
- [X] Method [preconditions](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [X] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [X] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests)
- [X] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/meta/API.html)
- [X] Change is documented in the [User Guide](http://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](http://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [ ] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

File: junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/discovery/DiscoverySelectorResolverTests.java [hunk 1]
```diff
@@ -287,16 +309,31 @@ void methodResolutionByUniqueIdWithParams() {
...
+		UniqueId uniqueId = uniqueIdForMethod(HerTestClass.class, "test7(java.math.BigDecimal)");
+		EngineDiscoveryRequest request = request().selectors(selectUniqueId(uniqueId)).build();
 
 		resolver.resolveSelectors(request, engineDescriptor);
 		assertTrue(engineDescriptor.getDescendants().isEmpty());
+		// @formatter:off
+		assertThat(listener.stream()
+				.filter(logRecord -> logRecord.getLevel() == Level.WARNING)
+				.filter(logRecord -> JavaElementsResolver.class.getName().equals(logRecord.getLoggerName()))
+				.map(LogRecord::getMessage)
```
Comment (sbrannen) [inline junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/discovery/DiscoverySelectorResolverTests.java]: You should be able to remove the two filters by using `org.junit.platform.commons.logging.LogRecordListener.stream(Class<?>, Level)`.

File: junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/discovery/DiscoverySelectorResolverTests.java [hunk 2]
```diff
@@ -287,16 +309,31 @@ void methodResolutionByUniqueIdWithParams() {
 		List<UniqueId> uniqueIds = uniqueIds();
 		assertThat(uniqueIds).contains(uniqueIdForClass(HerTestClass.class));
 		assertThat(uniqueIds).contains(uniqueIdForMethod(HerTestClass.class, "test7(java.lang.String)"));
+		// @formatter:off
+		assertThat(listener.stream()
+				.filter(logRecord -> JavaElementsResolver.class.getName().equals(logRecord.getLoggerName()))
+		).isEmpty();
```
Comment (sbrannen) [inline junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/discovery/DiscoverySelectorResolverTests.java]: You should be able to remove the filter by using `org.junit.platform.commons.logging.LogRecordListener.stream(Class<?>)`.

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/discovery/TestMethodResolver.java [hunk 3]
```diff
@@ -13,6 +13,7 @@
 import java.lang.reflect.Method;
 import java.util.function.Predicate;
 
+import org.junit.jupiter.api.Test;
```
Comment (sbrannen) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/discovery/TestMethodResolver.java]: Unused import?

File: junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/discovery/DiscoverySelectorResolverTests.java [hunk 4]
```diff
@@ -525,15 +562,82 @@ void testTemplateMethodResolutionByUniqueId() {
...
+		TestDescriptor testClassDescriptor = getOnlyElement(engineDescriptor.getChildren());
+
+		TestDescriptor testFactoryDescriptor = getOnlyElement(testClassDescriptor.getChildren());
+		DynamicDescendantFilter dynamicDescendantFilter = getDynamicDescendantFilter(testFactoryDescriptor);
+		assertThat(dynamicDescendantFilter.test(dynamicTestUid)).isTrue();
+		assertThat(dynamicDescendantFilter.test(differentDynamicTestUid)).isFalse();
+
+		// @formatter:off
+		assertThat(listener.stream()
+				.filter(logRecord -> JavaElementsResolver.class.getName().equals(logRecord.getLoggerName()))
```
Comment (sbrannen) [inline junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/discovery/DiscoverySelectorResolverTests.java]: ditto

File: junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/discovery/DiscoverySelectorResolverTests.java [hunk 5]
```diff
@@ -525,15 +562,82 @@ void testTemplateMethodResolutionByUniqueId() {
...
+
+		TestDescriptor testFactoryDescriptor = getOnlyElement(testClassDescriptor.getChildren());
+		DynamicDescendantFilter dynamicDescendantFilter = getDynamicDescendantFilter(testFactoryDescriptor);
+		assertThat(dynamicDescendantFilter.test(dynamicTestUid)).isTrue();
+		assertThat(dynamicDescendantFilter.test(differentDynamicContainerUid)).isFalse();
+		assertThat(dynamicDescendantFilter.test(differentDynamicTestUid)).isFalse();
+
+		// @formatter:off
+		assertThat(listener.stream()
+				.filter(logRecord -> JavaElementsResolver.class.getName().equals(logRecord.getLoggerName()))
```
Comment (sbrannen) [inline junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/discovery/DiscoverySelectorResolverTests.java]: ditto

**Comment by sbrannen** [comment]
Haven't had a chance to review in detail, but the overall direction looks very promising (if not already perfect 😇).

I've left a few minor notes in the interim.

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/discovery/TestMethodResolver.java (hunk 3 above)
Comment (marcphilipp) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/discovery/TestMethodResolver.java]: It's used in the Javadoc and was broken previously.

File: junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/discovery/DiscoverySelectorResolverTests.java (hunk 1 above)
Comment (marcphilipp) [reply junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/discovery/DiscoverySelectorResolverTests.java]: Very convenient 👍

File: junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/discovery/DiscoverySelectorResolverTests.java (hunk 2 above)
Comment (marcphilipp) [reply junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/discovery/DiscoverySelectorResolverTests.java]: 👍

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/discovery/TestMethodResolver.java (hunk 3 above)
Comment (sbrannen) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/discovery/TestMethodResolver.java]: ok

File: junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/discovery/DiscoverySelectorResolverTests.java (hunk 1 above)
Comment (sbrannen) [reply junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/discovery/DiscoverySelectorResolverTests.java]: glad you like it!

**Pull request merged**